Ms 40 view list and delete tracks

Henok-U opened the pull request.

Commit:
Merge branch 'dev' into MS-40-view-the-list-of-own-tracks

Commit:
Implement Track listing and deletion

- List all tracks owned by a user
- Delete a track owned by a user

Commit:
Test Written

Test for track viewing and deletion tested

Commit:
Fixed base_test.py

Commit:
Fixed bug in views.py documentation

Commit:
Update custom_user_schema.py

Removed Unused code

Commit:
Documentation views.py

Wrote documentation for Listing tracks and deleting a track

Commit:
Update custom_user_schema.py

Commit:
Update profile_view.py

removed redundant code

Commit:
Updated Documentation

- Restructuring code for swagger documentation

Commit:
Update profile_view.py

Fixed error

dmitry-sergeev commented on MusicHub/main/utils.py:
```diff
@@ -103,3 +108,29 @@ def create_or_return_user(backend, response, *args, **kwargs):
+    max_page_size = 40
+
+
+def format_to_minutes(sec):
```
`format_seconds_to_min` ?

dmitry-sergeev commented on MusicHub/main/utils.py:
```diff
@@ -103,3 +108,29 @@ def create_or_return_user(backend, response, *args, **kwargs):
+    return minutes + ":" + seconds
+
+
+def delete_generated_files_testing_tracks():
```
looks like it's needed only for the tests. If so let's move it to "test's utils"

dmitry-sergeev commented on MusicHub/tracks/custom_track_schema.py:
```diff
@@ -0,0 +1,50 @@
+class CustomAutoSchema(SwaggerAutoSchema):
+    def get_request_body_parameters(self, consumes):
+        result = super().get_request_body_parameters(consumes)
+        result.pop(0)
```
why we need it?

dmitry-sergeev commented on MusicHub/tracks/migrations/0003_alter_track_track_length.py:
```diff
@@ -0,0 +1,18 @@
+            name='track_length',
+            field=models.PositiveIntegerField(blank=True, null=True),
+        ),
+    ]
```
there are 3 migrations for the same app per one PR. Try to squash migrations for next PRs

dmitry-sergeev commented on MusicHub/tracks/models.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def get_upload_path(instance, filename):
+    return os.path.join(
```
`FileService` ?

dmitry-sergeev commented on MusicHub/tracks/models.py:
```diff
@@ -0,0 +1,39 @@
+    track_length = models.PositiveIntegerField(blank=True, null=True)
+    public = models.BooleanField(default=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+    created_by = models.ForeignKey(Common.AUTH_USER_MODEL, on_delete=models.CASCADE)
```
hm. I'm not sure that cascade deletion make sense here

dmitry-sergeev commented on MusicHub/tracks/serializers.py:
```diff
@@ -0,0 +1,50 @@
+        fields = ["filename", "track", "public"]
+
+    def validate_track(self, value):
+        if value.size > 30000000:
```
Fisrtly let's get rid of magic number. Than let's use format like `30 * 1024 * 1024` for the const `MAX_TRACK_SIZE_IN_MB`

dmitry-sergeev commented on MusicHub/tracks/serializers.py:
```diff
@@ -0,0 +1,50 @@
+        track = Track.objects.create(
+            created_by=self.context.get("user"), **validated_data
+        )
+        track.track_length = TinyTag.get(track.track.path).duration
```
I belive naming `file` instead of `track` in `Track` model better. For instance here `track.file.path` instead of `track.track.path`

dmitry-sergeev commented on MusicHub/tracks/serializers.py:
```diff
@@ -0,0 +1,50 @@
+    playlist = serializers.PrimaryKeyRelatedField(
+        many=False, read_only="True", default=None
+    )
+    # playlist = models.OneToOneField(Playlist, related_name='name')
```
remove or resolve

dmitry-sergeev commented on MusicHub/tracks/serializers.py:
```diff
@@ -0,0 +1,50 @@
+            "id": instance.id,
+            "filename": instance.filename,
+            "track_length": format_to_minutes(instance.track_length),
+            "created_at": instance.created_at.strftime("%d:%M:%Y"),
```
let's move it constants `"%d:%M:%Y"`

dmitry-sergeev commented on MusicHub/tracks/models.py:
```diff
@@ -0,0 +1,39 @@
+        validators=[FileExtensionValidator(["mp3", "wav", "aac"])],
+    )
+    track_length = models.PositiveIntegerField(blank=True, null=True)
+    public = models.BooleanField(default=True)
```
`is_public`

dmitry-sergeev commented on MusicHub/tracks/views.py:
```diff
@@ -0,0 +1,91 @@
+
+    permission_classes = [IsAuthenticated]
+    serializer_class = ListTrackSerializer
+    pagination_class = LargeResultsSetPagination
```
let's keep the following ordering:
1. attributes
2. functions

dmitry-sergeev commented on MusicHub/tracks/test/track_factory.py:
```diff
@@ -0,0 +1,20 @@
+    track_length = 145
+
+    @classmethod
+    def _create(cls, model_class, *args, **kwargs):
```
it looks weird because in order to use factory in standard way you have to override some "protected" methods (which starts from `_`)...

dmitry-sergeev commented on MusicHub/tracks/views.py:
```diff
@@ -0,0 +1,91 @@
+        )
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return Response(status=201, data=serializer.data)
```
again, why we need to reimplement DRF methods https://www.django-rest-framework.org/tutorial/3-class-based-views/#using-generic-class-based-views ?

dmitry-sergeev commented on MusicHub/users/views/profile_view.py:
```diff
@@ -91,6 +91,9 @@ class AddUpdateProfilePicture(GenericAPIView):
     """
 
+    def get_serializer(self, *args, **kwargs):
+        pass
```
?

HK116 committed:
Create .gitignore

HK116 committed:
Update utils.py and test_utils.py

- rename "format_to_minitues" to "format_sec_to_mins"
- moved "delete_generated_files_testing_tracks" to test utils.py

HK116 committed:
Squach migrations

HK116 committed:
Renaming and refactoring

HK116 committed:
Refactor test

- Removed redundant functions

HK116 committed:
Update views.py

- Removed unnecessarily overridden methods
- Refactored documentation

HK116 committed:
Renaming fields and methods

HK116 committed:
Update models.py

Replaced on_delete.CASCADE with on_delete.SET()

HK116 committed:
Update admin.py

- To show more relevant data on the admin page and to make searching easier

dmitry-sergeev commented on MusicHub/main/utils.py:
```diff
@@ -103,3 +104,26 @@ def create_or_return_user(backend, response, *args, **kwargs):
+            return minutes + ":" + seconds
+        except TypeError:
+            return "Invalid track length"
+    elif sec == None:
```
`not sec`

dmitry-sergeev commented on MusicHub/tracks/test/track_factory.py:
```diff
@@ -0,0 +1,20 @@
+        model = "tracks.Track"
+
+    id = factory.Faker("uuid4")
+    filename = "Track 01 - track"
```
email = factory.Sequence(lambda n: 'person{}@example.com'.format(n))

Henok-U committed:
Merge branch 'dev' into MS-40-view-the-list-of-own-tracks

HK116 committed:
Fixed errors

HK116 committed:
Update utils.py
